crates/provider/src/heart.rs

```diff
@@ -458,37 +503,112 @@ impl<S> Heartbeat<S> {
 }
 
 #[cfg(target_arch = "wasm32")]
-impl<S: Stream<Item = Block> + Unpin + 'static> Heartbeat<S> {
+impl<S: Stream<Item = Block> + Unpin + 'static, N: Network> Heartbeat<S, N> {
     /// Spawn the heartbeat task, returning a [`HeartbeatHandle`].
-    pub(crate) fn spawn(self) -> HeartbeatHandle {
+    pub(crate) fn spawn<T: Transport + Clone>(self, client: WeakClient<T>) -> HeartbeatHandle<N> {
         let (latest, latest_rx) = watch::channel(None::<Block>);
         let (ix_tx, ixns) = mpsc::channel(16);
 
-        self.into_future(latest, ixns).spawn_task();
+        self.into_future(latest, ixns, client).spawn_task();
 
         HeartbeatHandle { tx: ix_tx, latest: latest_rx }
     }
 }
 
 #[cfg(not(target_arch = "wasm32"))]
-impl<S: Stream<Item = Block> + Unpin + Send + 'static> Heartbeat<S> {
+impl<S: Stream<Item = Block> + Unpin + Send + 'static, N: Network> Heartbeat<S, N> {
     /// Spawn the heartbeat task, returning a [`HeartbeatHandle`].
-    pub(crate) fn spawn(self) -> HeartbeatHandle {
+    pub(crate) fn spawn<T: Transport + Clone>(self, client: WeakClient<T>) -> HeartbeatHandle<N> {
         let (latest, latest_rx) = watch::channel(None::<Block>);
         let (ix_tx, ixns) = mpsc::channel(16);
 
-        self.into_future(latest, ixns).spawn_task();
+        self.into_future(latest, ixns, client).spawn_task();
 
         HeartbeatHandle { tx: ix_tx, latest: latest_rx }
     }
 }
 
-impl<S: Stream<Item = Block> + Unpin + 'static> Heartbeat<S> {
-    async fn into_future(
+type FetchReceiptFuture<R> = Pin<Box<dyn Future<Output = Option<R>> + Send>>;
+
+struct ReceiptTasks<T, N: Network> {
+    client: WeakClient<T>,
+
+    /// Receiver of shutdown signal, used on unexpected channel close.
+    shutdown_tx: mpsc::Sender<()>,
+
+    /// Futures which perform a single request to fetch a receipt.
+    fetch_receipt_tasks: FuturesUnordered<FetchReceiptFuture<N::ReceiptResponse>>,
+
+    /// Futures which poll for receipts of confirmed transactions and notify [TxWatcher]s.
+    poll_receipt_tasks: FuturesUnordered<Pin<Box<dyn Future<Output = ()> + Send>>>,
+}
+
+impl<T: Transport + Clone, N: Network> ReceiptTasks<T, N> {
+    fn new(client: WeakClient<T>, shutdown_tx: mpsc::Sender<()>) -> Self {
+        Self {
+            client,
+            shutdown_tx,
+            fetch_receipt_tasks: Default::default(),
+            poll_receipt_tasks: Default::default(),
+        }
+    }
+
+    fn fetch_receipt(&mut self, tx_hash: B256) {
+        // We don't actually poll, but use poller to retry on failures.
```

Comment:
this is a bit of a layer violation, should be handled by a tower layer on the transport stack

Reply:
should this just do a request through client? what'd be the correct behavior on request failure?